remove stack usage reporting from the network stack

tyrantlizard77 opened the pull request:
This infrastructure is predicated on being able to dump out stack information with only a reference to the stack object, but this is unsafe. The only way to inspect a stack object safely requires information which is only found within its associated thread object.

Where it makes sence, I've replaced calls to the custom network stack logging functions with log_stack_usage(), ensuring that the thread name is set.

The "kernel stacks" command will dump out stack info for all threads. Stuff like dumping out stacks should be implemented once in the core kernel, and not on a per-subsystem basis.

Commit:
drivers: ieee802154: do not use net stack API

Use log_stack_usage() instead, which does the same thing,
safely.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: winc1500: do not use net_analyze_stack

This API is unsafe. Use log_stack_usage() instead and
set the thread name.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
net: purge NET_STACK and other stack APIs

The current design of the network-specific stack dumping APIs
is fundamentally unsafe. You cannot properly dump stack data
without information which is only available in the thread object.

In addition, this infrastructure is unnecessary. There is already
a core shell command which dumps stack information for all
active threads.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
samples: shell: enable stack info

Enable configs necessary for "kernel stacks" command.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
samples: pong: use log_stack_usage()

Don't use deprecated APIs.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: profiling_api: fix comments

This uses log_stack_usage() now.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
include: debug: remove old stack functions

These were unsafe.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
@jukkar I realize this is making a major change to the network stack but there's no way to get this right with the current approach, which uses stack objects without their threads.

I'd like to make the shell command "kernel stacks" be a suitable replacement, for all subsystems. Is dumping out information for all active threads + all interrupt stacks sufficient or do you need something beyond that for your debug use-cases?

jukkar commented on subsys/net/ip/net_shell.c:
```diff
@@ -3307,85 +3303,6 @@ static int cmd_net_route(const struct shell *shell, size_t argc, char *argv[])
-			     CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE);
-#endif
-
-static int cmd_net_stacks(const struct shell *shell, size_t argc,
```
Instead of fully removing this shell command, we could leave if for couple of releases in order to give the user a hint to enable kernel shell to get the same info. So I was thinking something like this:
```
static int cmd_net_stacks(const struct shell *shell, size_t argc,
			  char *argv[])
{
#if !defined(CONFIG_KERNEL_SHELL)
	PR("Enable CONFIG_KERNEL_SHELL and type \"kernel stacks\" to see stack information.\n");
#else 
	PR("Type \"kernel stacks\" to see stack information.\n");
#endif 
}
```

andrewboie replied:
sounds good, done.

jukkar reviewed (approved):
> I realize this is making a major change to the network stack but there's no way to get this right with the current approach, which uses stack objects without their threads

No worries, the net stacks information infrastructure was meant to disappear anyway eventually. It was there to get information about stacks when there was no similar thing in kernel side. I had plans to remove it too eventually so +1 to this.

Commit:
shell: show IRQ stack information

This is placeholder code; better kernel support for dumping
exception/interrupt related stacks is forthcoming.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
